examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+            let state = host.state_mut();
+            ensure!(
+                sender == params.owner || state.is_operator(&sender, &params.owner),
+                ContractError::Unauthorized
+            );
+
+            // Update the state.
+            state.burn(&TOKEN_ID_WCCD, params.amount, &params.owner)?;
+
+            // Log the burning of tokens.
+            logger.log(&Cis2Event::Burn(BurnEvent {
+                token_id: TOKEN_ID_WCCD,
+                amount:   params.amount,
+                owner:    params.owner,
+            }))?;
+
+            let unwrapped_amount = Amount::from_micro_ccd(params.amount.into());
+
+            match params.receiver {
+                Receiver::Account(address) => host.invoke_transfer(&address, unwrapped_amount)?,
+                Receiver::Contract(address, function) => {
+                    host.invoke_contract(
+                        &address,
+                        &params.data,
+                        function.as_entrypoint_name(),
+                        unwrapped_amount,
+                    )?;
+                }
+            }
+
+            Ok(())
+        }
+    }
+}
+
+// Contract functions required by CIS2
+
+#[allow(dead_code)]
+type TransferParameter = TransferParams<ContractTokenId, ContractTokenAmount>;
+
+/// Execute a list of token transfers, in the order of the list.
+///
+/// Logs a `Transfer` event and invokes a receive hook function for every
+/// transfer in the list.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Any of the transfers fail to be executed, which could be if:
+///     - The `token_id` does not exist.
+///     - The sender is not the owner of the token, or an operator for this
+///       specific `token_id` and `from` address.
+///     - The token is not owned by the `from`.
+/// - Fails to log event.
+/// - Any of the receive hook function calls rejects.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "transfer",
+    parameter = "TransferParameter",
+    enable_logger,
+    mutable
+)]
+fn contract_transfer<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    let slot_time = ctx.metadata().slot_time();
+
+    match host.state().state_address {
+        None => bail!(concordium_cis2::Cis2Error::Custom(OnlyState)),
+
+        Some(state_address) => {
+            let unpause_time = host
+                .invoke_contract_raw(
+                    &state_address,
+                    Parameter(&[]),
+                    EntrypointName::new_unchecked("get_unpause_time"),
+                    Amount::zero(),
+                )?
+                .1;
+
+            let unpause_time = if let Some(mut unpause_time) = unpause_time {
+                unpause_time.get()?
+            } else {
+                return Err(concordium_cis2::Cis2Error::Custom(StateInvokeError));
+            };
```

Comment:
This repeats a lot, so it would probably make sense to have a function for it if it stays like this.